File: libsrc/pylith/meshio/OutputTriggerStep.cc
// =================================================================================================
// This code is part of PyLith, developed through the Computational Infrastructure
// for Geodynamics (https://github.com/geodynamics/pylith).
//
// Copyright (c) 2010-2025, University of California, Davis and the PyLith Development Team.
// All rights reserved.
//
// See https://mit-license.org/ and LICENSE.md and for license information.
// =================================================================================================

#include <portinfo>

#include "pylith/meshio/OutputTriggerStep.hh" // Implementation of class methods

#include "pylith/utils/constants.hh" // USES pylith::min_int
#include "pylith/utils/error.hh" // USES PYLITH_METHOD_BEGIN/END
#include "pylith/utils/journals.hh" // USES PYLITH_COMPONENT_*

// ---------------------------------------------------------------------------------------------------------------------
// Constructor
pylith::meshio::OutputTriggerStep::OutputTriggerStep(void) :
    _numStepsSkip(0),
    _stepWrote(pylith::min_int+10) { // constructor
    PyreComponent::setName("outputtriggerstep");
} // constructor


// ---------------------------------------------------------------------------------------------------------------------
// Destructor
pylith::meshio::OutputTriggerStep::~OutputTriggerStep(void) {}


// ---------------------------------------------------------------------------------------------------------------------
// Set number of steps to skip between writes.
void
pylith::meshio::OutputTriggerStep::setNumStepsSkip(const int value) {
    PYLITH_COMPONENT_DEBUG("OutputTriggerStep::setNumStepsSkip(value="<<value<<")");

    _numStepsSkip = (value >= 0) ? value : 0;
} // setNumStepsSkip


// ---------------------------------------------------------------------------------------------------------------------
// Get number of steps to skip between writes.
int
pylith::meshio::OutputTriggerStep::getNumStepsSkip(void) const {
    return _numStepsSkip;
} // getNumStepsSkip


// ---------------------------------------------------------------------------------------------------------------------
// Check whether we want to write output at time t.
bool
pylith::meshio::OutputTriggerStep::shouldWrite(const PylithReal t,
                                               const PylithInt tindex) {
    PYLITH_METHOD_BEGIN;
    PYLITH_COMPONENT_DEBUG("OutputTriggerStep::shouldWrite(t="<<t<<", tindex="<<tindex<<")");

    bool isWrite = false;
    if (tindex - _stepWrote > _numStepsSkip) {
        isWrite = true;
        _stepWrote = tindex;
    } // if

    PYLITH_METHOD_RETURN(isWrite);
} // shouldWrite


// End of file

File: pylith/meshio/OutputTriggerStep.py
# =================================================================================================
# This code is part of PyLith, developed through the Computational Infrastructure
# for Geodynamics (https://github.com/geodynamics/pylith).
#
# Copyright (c) 2010-2023, University of California, Davis and the PyLith Development Team.
# All rights reserved.
#
# See https://mit-license.org/ and LICENSE.md and for license information. 
# =================================================================================================

from .OutputTrigger import OutputTrigger
from .meshio import OutputTriggerStep as ModuleOutputTriggerStep


class OutputTriggerStep(OutputTrigger, ModuleOutputTriggerStep):
    """
    Define how often output is written in terms of solution steps.

    Implements `OutputTrigger`.
    """
    DOC_CONFIG = {
        "cfg": """
            [output_trigger]
            num_skip = 2
        """
    }

    import pythia.pyre.inventory

    numSkip = pythia.pyre.inventory.int("num_skip", default=0, validator=pythia.pyre.inventory.greaterEqual(0))
    numSkip.meta['tip'] = "Number of solution steps to skip between writes (0 means write every time step)."

    def __init__(self, name="outputtriggerstep"):
        """Constructor.
        """
        OutputTrigger.__init__(self, name)

    def preinitialize(self):
        """Setup output trigger.
        """
        ModuleOutputTriggerStep.__init__(self)
        ModuleOutputTriggerStep.setIdentifier(self, self.aliases[-1])
        ModuleOutputTriggerStep.setNumStepsSkip(self, self.numSkip)

    def _configure(self):
        """Set members based using inventory.
        """
        OutputTrigger._configure(self)

# FACTORIES ////////////////////////////////////////////////////////////


def output_trigger():
    """Factory associated with OutputTriggerStep.
    """
    return OutputTriggerStep()


# End of file
